Mathlib/Data/Nat/NormNum.lean

```diff
@@ -0,0 +1,251 @@
+  | a, _, ⟨rfl⟩, h => ⟨by rw [cast_ofNat, minFac_eq_two_iff, Nat.dvd_iff_mod_eq_zero, h]⟩
+
+theorem isNat_minFac_3 : {n : ℕ} → {n' : ℕ} → {k : ℕ} → {k' : ℕ} →
+    IsNat n n' → IsNat k k' → MinFacHelper n' k → n' % k = 0 → IsNat (minFac n) k'
+  | n, _, k, _, ⟨rfl⟩, ⟨rfl⟩, h1, h2 => by
+    rw [← Nat.dvd_iff_mod_eq_zero] at h2
+    exact ⟨le_antisymm (minFac_le_of_dvd h1.1.le h2) h1.2.2⟩
+
+theorem isNat_minFac_4 : {n : ℕ} → {n' : ℕ} → {k : ℕ} →
+    IsNat n n' → MinFacHelper n' k → n' < k * k → IsNat (minFac n) n'
+  | n, _, k, ⟨rfl⟩, h1, h2 => by
+    refine ⟨(Nat.prime_def_minFac.mp ?_).2⟩
+    rw [Nat.prime_def_le_sqrt]
+    refine ⟨h1.one_lt, λ m hm hmn h2mn ↦ ?_⟩
+    exact lt_irrefl m <| calc
+      m ≤ sqrt n   := hmn
+      _ < k        := sqrt_lt.mpr h2
+      _ ≤ n.minFac := h1.2.2
+      _ ≤ m        := Nat.minFac_le_of_dvd hm h2mn
+
+theorem not_prime_of_lt_two (h : n < 2) : ¬ n.Prime :=
+  fun hn => hn.two_le.not_lt h
+
+/-- A partial proof of `factors`. Asserts that `l` is a sorted list of primes, lower bounded by a
+prime `p`, which multiplies to `n`. -/
+def FactorsHelper (n p : ℕ) (l : List ℕ) : Prop :=
+  p.Prime → List.Chain (· ≤ ·) p l ∧ (∀ a ∈ l, Nat.Prime a) ∧ List.prod l = n
+
+theorem factorsHelper_nil (a : ℕ) : FactorsHelper 1 a [] := fun _ =>
+  ⟨List.Chain.nil, by rintro _ ⟨⟩, List.prod_nil⟩
+
+theorem factorsHelper_cons' (n m a b : ℕ) (l : List ℕ) (h₁ : b * m = n) (h₂ : a ≤ b)
+    (h₃ : minFac b = b) (H : FactorsHelper m b l) : FactorsHelper n a (b :: l) := fun pa =>
+  have pb : b.Prime := Nat.prime_def_minFac.2 ⟨le_trans pa.two_le h₂, h₃⟩
+  let ⟨f₁, f₂, f₃⟩ := H pb
+  ⟨List.Chain.cons h₂ f₁,
+    fun c h => (List.eq_or_mem_of_mem_cons h).elim (fun e => e.symm ▸ pb) (f₂ _),
+    by rw [List.prod_cons, f₃, h₁]⟩
+
+theorem factorsHelper_cons (n m a b : ℕ) (l : List ℕ) (h₁ : b * m = n) (h₂ : a < b)
+    (h₃ : minFac b = b) (H : FactorsHelper m b l) : FactorsHelper n a (b :: l) :=
+  factorsHelper_cons' _ _ _ _ _ h₁ h₂.le h₃ H
+
+theorem factorsHelper_sn (n a : ℕ) (h₁ : a < n) (h₂ : minFac n = n) : FactorsHelper n a [n] :=
+  factorsHelper_cons _ _ _ _ _ (mul_one _) h₁ h₂ (factorsHelper_nil _)
+
+theorem factorsHelper_same (n m a : ℕ) (l : List ℕ) (h : a * m = n) (H : FactorsHelper m a l) :
+    FactorsHelper n a (a :: l) := fun pa =>
+  factorsHelper_cons' _ _ _ _ _ h le_rfl (Nat.prime_def_minFac.1 pa).2 H pa
+
+theorem factorsHelper_same_sn (a : ℕ) : FactorsHelper a a [a] :=
+  factorsHelper_same _ _ _ _ (mul_one _) (factorsHelper_nil _)
+
+theorem factorsHelper_end (n : ℕ) (l : List ℕ) (H : FactorsHelper n 2 l) : Nat.factors n = l :=
+  let ⟨h₁, h₂, h₃⟩ := H Nat.prime_two
+  have := List.chain'_iff_pairwise.1 (@List.Chain'.tail _ _ (_ :: _) h₁)
+  (List.eq_of_perm_of_sorted (Nat.factors_unique h₃ h₂) this (Nat.factors_sorted _)).symm
+
+open NormNum
+
+/-- Produce a proof that `n` is not prime from a factor `1 < d < n`. -/
+def deriveNotPrime (n d : ℕ) : MetaM <| Q(¬ Nat.Prime $n) := do
```

Comment:
```suggestion
def deriveNotPrime (n d : ℕ) : Q(¬ Nat.Prime $n) := do
```
ditto for `evalMinFac.aux`. If you are lucky this might fix it so it becomes tail recursive

Reply:
Oh, that is indeed nicer.

I still get a deep recursion when running it on primes roughly 2 ^ 28. Shall I turn `evalMinFac.aux` into a `for` loop?